cmd/pod-scaler/producer.go

```diff
@@ -0,0 +1,482 @@
+		c.lock.RLock()
+		numSteps := c.maxSize - 1
+		c.lock.RUnlock()
+		for {
+			if start == uncoveredRange.End {
+				break
+			}
+			if int64(stop.Sub(start)/r.Step) > numSteps {
+				stop = start.Add(time.Duration(numSteps) * r.Step)
+			}
+			c.wg.Add(1)
+			go q.executeOverRange(ctx, c, prometheusapi.Range{Start: start, End: stop, Step: r.Step})
+			start = stop
+			stop = uncoveredRange.End
+		}
+	}
+	c.wg.Wait()
+	logger.Infof("Query completed after %s.", time.Since(queryStart))
+	close(c.errors)
+	errLock.Lock()
+	return kerrors.NewAggregate(errs)
+}
+
+func (q *querier) executeOverRange(ctx context.Context, c *clusterMetadata, r prometheusapi.Range) {
+	defer c.wg.Done()
+	numSteps := int64(r.End.Sub(r.Start) / r.Step)
+	logger := c.logger.WithFields(logrus.Fields{
+		"start": r.Start.Format(time.RFC3339),
+		"end":   r.End.Format(time.RFC3339),
+		"step":  r.Step,
+		"steps": numSteps,
+	})
+	if err := c.sync.Acquire(ctx, numSteps); err != nil {
+		c.errors <- err
+		return
+	}
+	defer c.sync.Release(numSteps)
+	c.lock.RLock()
+	currentMax := c.maxSize
+	c.lock.RUnlock()
+	subdivide := func() {
+		c.wg.Add(2)
+		middle := r.Start.Add(time.Duration(numSteps) / 2 * r.Step)
+		go q.executeOverRange(ctx, c, prometheusapi.Range{Start: r.Start, End: middle, Step: r.Step})
+		go q.executeOverRange(ctx, c, prometheusapi.Range{Start: middle, End: r.End, Step: r.Step})
+	}
+	if numSteps >= currentMax {
+		logger.Debugf("Preemptively halving request as prior data shows ours is too large (%d>=%d).", numSteps, currentMax)
+		subdivide()
+		return
+	}
+
+	queryStart := time.Now()
+	logger.Debug("Querying Prometheus.")
+	q.lock.RLock()
+	metric := q.data.Metric
+	q.lock.RUnlock()
+	result, warnings, err := c.client.QueryRange(ctx, queryFor(metric), r)
+	logger.Debugf("Queried Prometheus API in %s.", time.Since(queryStart))
+	if err != nil {
+		apiError := &prometheusapi.Error{}
+		if errors.As(err, &apiError) {
+			// Prometheus determined not to expose this programmatically ...
+			if strings.HasSuffix(apiError.Msg, "504") {
+				var exitEarly bool
+				c.lock.Lock()
+				if numSteps < c.maxSize {
+					if numSteps > 250 {
+						// We hit a timeout and are still asking for a reasonably "large" amount of
+						// data at once, so halve the amount of data we are asking for in order to
+						// have a higher chance of getting the data next time. If we're asking for
+						// a small amount already it's likely the server is on the verge of falling
+						// over, so just error out and try again later.
+						logger.Debugf("Received 504 asking for %d samples, halving to %d.", numSteps, numSteps/2)
+						c.maxSize = numSteps
+						exitEarly = true
+					} else {
+						logger.Debugf("Received 504 but only asking for %d samples, aborting.", numSteps)
+					}
+				} else {
+					// We hit a timeout asking for a known large value,
+					exitEarly = true
+				}
+				c.lock.Unlock()
+				if exitEarly {
+					subdivide()
+					return
+				}
+			}
+		}
```

Comment:
This is quite hard to read. I played with it a bit in an editor and I think the flow can be simplified like this to reduce the amount of branching and indentation (proposing, I don't insist). Could be even more straightforward using a switch with fallthrough, but I just hate fallthrough in Go.

```go
		// Prometheus determined not to expose this programmatically ...
		if errors.As(err, &apiError) && strings.HasSuffix(apiError.Msg, "504") {
			c.lock.Lock()
			divide := func() { c.lock.Unlock(); subdivide() }
			if numSteps >= c.maxSize {
				// We hit a timeout asking for a known large value,
				divide()
				return
			}
			if numSteps > 250 {
				// We hit a timeout and are still asking for a reasonably "large" amount of
				// data at once, so halve the amount of data we are asking for in order to
				// have a higher chance of getting the data next time. If we're asking for
				// a small amount already it's likely the server is on the verge of falling
				// over, so just error out and try again later.
				logger.Debugf("Received 504 asking for %d samples, halving to %d.", numSteps, numSteps/2)
				c.maxSize = numSteps
				divide()
				return
			}
			logger.Debugf("Received 504 but only asking for %d samples, aborting.", numSteps)
			c.lock.Unlock()
		}
```

Reply:
I think I went the way I did because there is one and only one return, which means you can more simply reason about locking and unlocking the critical section. I'll play around with making this `if` less nested though.